timemachine/fe/rest/interpolation.py

```diff
@@ -0,0 +1,106 @@
+from dataclasses import dataclass
+from typing import Generic, Protocol, TypeVar
+
+import matplotlib.pyplot as plt
+import numpy as np
+from numpy.typing import ArrayLike, NDArray
+
+
+class InterpolationFxn(Protocol):
+    @property
+    def src(self) -> ArrayLike:
+        ...
+
+    @property
+    def dst(self) -> ArrayLike:
+        ...
+
+    def __call__(self, x: ArrayLike) -> NDArray:
+        ...
+
+
+@dataclass(frozen=True)
+class Linear:
+    src: ArrayLike
+    dst: ArrayLike
+
+    def __call__(self, x: ArrayLike) -> NDArray:
+        src = np.asarray(self.src)
+        dst = np.asarray(self.dst)
+        x = np.asarray(x)
+
+        a = dst - src
+        b = src
+        return a * x + b
+
+
+@dataclass(frozen=True)
+class Quadratic:
+    src: ArrayLike
+    dst: ArrayLike
+
+    def __post_init__(self):
+        assert np.all(self.src != self.dst)
+
+    def __call__(self, x: ArrayLike) -> NDArray:
+        src = np.asarray(self.src)
+        dst = np.asarray(self.dst)
+        x = np.asarray(x)
+
+        a = src + dst - 2.0 * np.sqrt(src * dst)
+        b = (src + np.sqrt(src * dst)) / (dst - src)
+
+        # special-case 0 and 1 to avoid issues with roundoff error
+        return np.where(
+            x == 0.0,
+            self.src,
+            np.where(
+                x == 1.0,
+                self.dst,
+                a * (x + b) ** 2,
+            ),
+        )
+
+
+@dataclass(frozen=True)
+class Exponential:
+    src: ArrayLike
+    dst: ArrayLike
+
+    def __call__(self, x: ArrayLike) -> NDArray:
+        src = np.asarray(self.src)
+        dst = np.asarray(self.dst)
+        x = np.asarray(x)
+
+        a = src
+        b = np.log(dst / src)
+        return a * np.exp(b * x)
+
+
+F = TypeVar("F", bound=InterpolationFxn)
+
+
+@dataclass(frozen=True)
+class Symmetric(Generic[F]):
+    f: F
+
+    @property
+    def src(self):
+        return self.f.src
+
+    @property
+    def dst(self):
+        return self.f.src
+
+    def __call__(self, x: ArrayLike) -> NDArray:
+        x = np.asarray(x)
+        return np.where(
+            x < 0.5,
+            self.f(2.0 * x),
+            self.f(2.0 * (1.0 - x)),
+        )
```

Comment:
Good call; the definitions in `fe.rest.interpolation` all implicitly assume a domain of [0, 1], but this was not validated. Added a check in https://github.com/proteneer/timemachine/pull/1424/commits/4ab60753bdd46e91456b03ca339c90692fdfb421.

Reply:
Unfortunately the assertions broke compositionality in `Symmetric`, because `np.where` evaluates both branches of the piecewise function. Thinking about this more, it seems like the fundamental issue was that previously it was unclear that the definitions in `fe.rest.interpolate` were specialized to a domain of [0, 1]. I've tried to clarify this with module and class docstrings in https://github.com/proteneer/timemachine/pull/1424/commits/8b9332bbbe6014d38a472b719921ccea7aabf6ec.